File: packages/mp_dom_runtime/src/components/mpkit/platform_view.ts
import { ComponentView } from "../component_view";

export class MPPlatformView extends ComponentView {
  private static _invokeMethodCompleter: { [key: string]: (result: any) => void } = {};

  static handleInvokeMethodCallback(seqId: string, result: any) {
    if (this._invokeMethodCompleter[seqId]) {
      this._invokeMethodCompleter[seqId](result);
      delete this._invokeMethodCompleter[seqId];
    }
  }

  classname = "PlatformView";
  _currentSize?: { width: number; height: number };

  constructor(readonly document: Document, readonly initialAttributes?: any) {
    super(document, initialAttributes);
    if (this.elementType().indexOf(".") > 0 && __MP_TARGET_BROWSER__) {
      this.htmlElement = this.createFromWebTemplate();
    } else if (initialAttributes?.tag && __MP_TARGET_BROWSER__) {
      this.htmlElement = document.createElement(initialAttributes.tag);
    }
    if (__MP_MINI_PROGRAM__) {
      this.htmlElement.getBoundingClientRect = (this.htmlElement as any).$$getBoundingClientRect;
    }
    this.htmlElement.id = "platform_view_" + Math.random().toString();
    this.htmlElement.setAttribute("id", this.htmlElement.id);
  }

  elementType() {
    return "div";
  }

  setSize(size: { width: number; height: number }) {
    if (this._currentSize && this._currentSize.width === size.width && this._currentSize.height === size.height) return;
    this._currentSize = size;
    this.engine.sendMessage(
      JSON.stringify({
        type: "platform_view",
        message: {
          event: "setSize",
          hashCode: this.hashCode,
          size,
        },
      })
    );
  }

  createFromWebTemplate(): HTMLElement {
    if (!__MP_TARGET_BROWSER__) return null!;
    let templateNode = document.getElementsByName(this.elementType())[0] as HTMLTemplateElement;
    if (!templateNode) {
      throw `Template ${this.elementType()} not found.`;
    }
    let clone = document.importNode(templateNode.content, true);
    return clone.children[0] as HTMLElement;
  }

  setAttributes(attributes: any): void {
    super.setAttributes(attributes);
    if (__MP_TARGET_BROWSER__) {
      let keys: string[] = [];
      try {
        keys = attributes.__keys__ ?? Object.keys(attributes);
      } catch (error) {
        keys = Object.keys(attributes);
      }
      keys.forEach((key) => {
        try {
          this.htmlElement.setAttribute(key, attributes[key]);
        } catch (error) {
          console.error(error);
        }
      });
    }
  }

  onMethodCall(method: string, params: any) {}

  invokeMethod(method: string, params: any, requireResult: boolean = false): Promise<any> | undefined {
    let seqId = `${this.hashCode}_${Math.random()}`;
    this.engine.sendMessage(
      JSON.stringify({
        type: "platform_view",
        message: {
          event: "methodCall",
          hashCode: this.hashCode,
          method,
          params,
          seqId,
          requireResult,
        },
      })
    );
    if (requireResult) {
      return new Promise((res) => {
        MPPlatformView._invokeMethodCompleter[seqId] = res;
      });
    }
  }
}

File: packages/mp_flutter_runtime/lib/components/mpkit/platform_view.dart
part of '../../mp_flutter_runtime.dart';

class MPPlatformView extends ComponentView {
  static final Map<String, Completer> _invokeMethodCallback = {};

  static void _didReceivedPlatformViewMessage(Map data, MPEngine engine) async {
    String? event = data['event'];
    if (event == 'methodCall') {
      int? hashCode = data['hashCode'];
      if (hashCode != null) {
        final target = engine._componentFactory._cacheViews[hashCode]?.widget;
        if (target is MPPlatformView) {
          final result = await target.onMethodCall(
            data['method'],
            data['params'],
          );
          if (data['requireResult'] == true) {
            engine._sendMessage({
              'type': 'platform_view',
              'message': {
                'event': 'methodCallCallback',
                'seqId': data['seqId'],
                'result': result,
              },
            });
          }
        }
      }
    } else if (event == 'methodCallCallback') {
      String? seqId = data['seqId'];
      if (seqId != null) {
        _invokeMethodCallback[seqId]?.complete(data['result']);
        _invokeMethodCallback.remove(seqId);
      }
    }
  }

  MPPlatformView({
    Key? key,
    Map? data,
    Map? parentData,
    required _MPComponentFactory componentFactory,
  }) : super(
            key: key,
            data: data,
            parentData: parentData,
            componentFactory: componentFactory);

  Future onMethodCall(String method, dynamic params) async {
    return null;
  }

  void invokeMethod(String method, dynamic params) {
    if (dataHashCode == null) return;
    final engine = componentFactory.engine;
    final seqId = Random().nextDouble().toString();
    engine._sendMessage({
      'type': 'platform_view',
      'message': {
        'event': 'methodCall',
        'hashCode': dataHashCode,
        'method': method,
        'params': params,
        'seqId': seqId,
      },
    });
  }

  Future invokeMethodWithResult(String method, dynamic params) async {
    if (dataHashCode == null) return;
    final engine = componentFactory.engine;
    final seqId = Random().nextDouble().toString();
    final completer = Completer();
    _invokeMethodCallback[seqId] = completer;
    engine._sendMessage({
      'type': 'platform_view',
      'message': {
        'event': 'methodCall',
        'hashCode': dataHashCode,
        'method': method,
        'params': params,
        'seqId': seqId,
        'requireResult': true,
      },
    });
    return completer.future;
  }

  @override
  Widget builder(BuildContext context) {
    return const SizedBox();
  }
}
